app/models/article.rb

```diff
@@ -554,6 +554,7 @@ def set_tag_list(tags)
     self.tag_list = [] # overwrite any existing tag with those from the front matter
     tag_list.add(tags, parse: true)
     self.tag_list = tag_list.map { |tag| Tag.find_preferred_alias_for(tag) }
+    self.cached_tag_list = tag_list.join(", ")
```

Comment:
it looks like modifying the cached tag list here breaks the assertion on validation 

Article spec is failing to prove this:

```ruby
    it { is_expected.to validate_length_of(:cached_tag_list).is_at_most(126) }
```

The behavior suggests that the before_validation evaluation of body markdown is now modifying the cached tag list, so setting it to an invalid length, and expecting an error message, fails:

```ruby
article.cached_tag_list
=> "javascript, html, discuss"                                                  
article.cached_tag_list = (article.tag_list * 20).join(', ') ;  
article.cached_tag_list.length
=> 538
article.valid?
=> true
article.cached_tag_list
=> "javascript, html, discuss"                                                  
```

This breaks the "validate_length_of" check by changing the value rather than raising the expected error.

Reply:
Is the length for the tag_list enforced at the database level?  That's one reason I could see keeping this.

(Also…I would say that it's more robust to so `30 * 4 + (4-1) * 2` (where 2 is the length of `, `)).  But who knows if we're going to nudge that 4 to a 5 or what have you.

Note, none of this quite addresses the issue, but pokes around the edges a bit more.

Reply:
the column type is string - which I think may default to varchar(128) on some systems - and the length restriction of 30 on tag names was _possibly_ imposed by this requirement rather than the other direction (because we can only fit so much in the serialized tag name list, restrict the length of tag names to fit) unless it's a front end formatting concern (long tags are hard to show in the space allotted).

The create table statement for articles just says "character varying" without specifying a max length in the db:

```sql
CREATE TABLE public.articles (
    id bigint NOT NULL,
    ...
    cached_organization text,
    cached_tag_list character varying,
    cached_user text,
    cached_user_name character varying,
    cached_user_username character varying,
    ...

--- there is also an index on the list for faster search:
CREATE INDEX index_articles_on_cached_tag_list ON public.articles USING gin (cached_tag_list public.gin_trgm_ops);
```

It looks like 126 might also be a magic number within postgresql for string columns: 

https://www.postgresql.org/docs/13/datatype-character.html

>  The storage requirement for a short string (up to 126 bytes) is 1 byte plus the actual string, which includes the space padding in the case of character.

Reply:
I wonder if this has to do with the column being `String` instead of `Text`. And/or, there are many quirks with this gem and caching. In your example, I'm not sure that code is _actually_ updating the `cached_tag_list` the way we think. I vaguely remember running into this before: https://app.gitbook.com/o/-MEieVR_Ti0SWAGatoMl/s/-MEieZK0QgGoR2nm6E1D-2910905616/engineering/tech-overviews/acts-as-taggable

Reply:
Ha, same thought @djuber wrt column type. Re-reading what I wrote way back...I think we may want to change how we set this and use `tag_list`.

Reply:
It doesn't look like there's a db level length constraint on this column

```
Forem_development=# update articles set cached_tag_list = 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA, AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA, AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA, AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA' where id = 19;
UPDATE 1

Forem_development=# select id, length(cached_tag_list) from articles where id = 19;
 id | length 
----+--------
 19 |    246
(25 rows)
```

Reply:
I see. Super thankful for the context around `tag_list` @atsmith813

I'll check if the issue is fixable by having `Follows::UpdatePointsWorker` refer to the `tag_list`, and not the `cached_tag_list`. It's a much smaller change, farther from a core model class, and makes me feel easier to contemplate the consequences.

Reply:
Looking closer at the factory created article and the db sequence just to confirm what's happening:

```
 Article Create (3.8ms)  INSERT INTO "articles" ("body_markdown", "cached_user", "cached_user_name", "cached_user_username", "created_at", "description", "experience_level_rating", "featured_number", "last_comment_at", "main_image", "nth_published_by_author", "password", "path", "processed_html", "published", "published_at", "reading_time", "slug", "title", "updated_at", "user_id") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21) RETURNING "id" 
 ```

When we insert the article the first time, the tag list is in the markdown body, but the taggings don't exist yet (they'll be added immediately after to the taggins join table).

```
   ActsAsTaggableOn::Tagging Create (1.4ms)  INSERT INTO "taggings" ("context", "created_at", "tag_id", "taggable_id", "taggable_type") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["context", "tags"], ["created_at", "2021-11-22 23:05:29.996016"], ["tag_id", 1737], ["taggable_id", 1021], ["taggable_type", "Article"]]
  ActsAsTaggableOn::Tag Update All (0.6ms)  UPDATE "tags" SET "taggings_count" = COALESCE("taggings_count", 0) + $1 WHERE "tags"."id" = $2  [["taggings_count", 1], ["id", 1737]]
  User Update (0.7ms)  UPDATE "users" SET "updated_at" = $1, "latest_article_updated_at" = $2 WHERE "users"."id" = $3  [["updated_at", "2021-11-22 23:05:30.003322"], ["latest_article_updated_at", "2021-11-22 23:05:30.003322"], ["id", 1094]]
```

Then reloading the article and saving it triggers setting the cached column:

```
  Article Update (1.0ms)  UPDATE "articles" SET "cached_tag_list" = $1, "updated_at" = $2 WHERE "articles"."id" = $3  [["cached_tag_list", "tag"], ["updated_at", "2021-11-22 23:14:23.313125"], ["id", 1022]]
```